File: 2024/05_printing_queue.py
input = """47|53
97|13
97|61
97|47
75|29
61|13
75|53
29|13
97|29
53|29
61|53
97|53
61|29
47|13
75|47
97|75
47|61
75|61
47|29
75|13
53|13

75,47,61,53,29
97,61,53,29,13
75,29,13
75,97,47,61,53
61,13,29
97,13,75,29,47"""

input = open('data/05.dat').read()


rules = []
updates = []
for line in input.splitlines():
    if '|' in line:
        rules.append(list(map(int, line.split('|'))))
    elif ',' in line:
        updates.append(list(map(int, line.split(','))))


def is_correctly_ordered(update, rules):
    for rule in rules:
        left, right = rule
        if left in update and right in update:
            if not (update.index(left) < update.index(right)):
                return False
    return True


def get_middle(update):
    return update[len(update) // 2]


print("Part 1: ", sum(get_middle(update) for update in updates if is_correctly_ordered(update, rules)))


def order_pages(pages):
    # Create the applicable rule set:
    local_rule_set = []
    for rule in rules:
        left, right = rule
        if left in pages and right in pages:
            local_rule_set.append(rule)

    while not is_correctly_ordered(pages, local_rule_set):
        for rule in local_rule_set:
            left, right = rule
            left_at, right_at = pages.index(left), pages.index(right)
            while left_at > right_at:
                _ = pages.pop(pages.index(right))
                pages.insert(left_at + 1, right)
                left_at, right_at = pages.index(left), pages.index(right)

    return pages

print("Part 2: ", sum(get_middle(order_pages(update)) for update in updates if not is_correctly_ordered(update, rules)))

File: 2024/05_printing_queue.jl
rules, queues = let
    lines = readlines("data/05.dat")

    rules_active = true
    r, q = Any[], Any[]
    for line in lines
        if length(line) == 0
            rules_active = false
            continue
        end

        if rules_active
            left, right = [parse(Int, x) for x in split(strip(line, '\n'), '|')]
            push!(r, (left, right))
        else
            push!(q, [parse(Int, x) for x in split(strip(line, '\n'), ',')])
        end
    end
    
    r, q
end


function score_queue(queue, rules)
    for (left, right) in rules
        if left in queue && right in queue
            lidx, ridx = findfirst(x->x == left, queue), findfirst(x->x == right, queue)
            if lidx >= ridx
                return 0
            end
        end
    end
    return queue[div((length(queue) - 1), 2) + 1]

    #= Alternative version which tries sub-lists. Definitely less readable but for shorter lists should be faster
    for k in range(2, length(queue))
        must_not_exist_left = queue[1:k-1]
        q = queue[k]
        if q in keys(rules)
            if rules[q] in must_not_exist_left
                return 0
            end
        end
    end
    return queue[div((length(queue) - 1), 2) + 1]
    =#
end


function order_pages(queue, rules)
    # Create the applicable, local rule set:
    applicable_rules = [(left, right) for (left, right) in rules if left in queue && right in queue]

    while score_queue(queue, rules) == 0
        for (left, right) in applicable_rules
            lidx, ridx = findfirst(x->x==left, queue), findfirst(x->x==right, queue)
            if lidx > ridx
                popat!(queue, ridx)
                insert!(queue, lidx, right)
            end
        end
    end

    return score_queue(queue, rules)
end


@time begin
    println("Part 1: ", sum([score_queue(queue, rules) for queue in queues]))
    println("Part 2: ", sum([order_pages(queue, rules) for queue in queues if score_queue(queue, rules) == 0]))
end
